Allow negative `n` in `slice_head` and `slice_tail`

eutwt opened the pull request:
Closes #5961

romainfrancois commented on tests/testthat/test-slice.r:
```diff
@@ -189,11 +189,23 @@ test_that("slice() handles matrix and data frame columns (#3630)", {
+test_that("slice_head/slice_tail work with negative n", {
+  res <- slice_head(mtcars, n = -2)
+  exp <- head(mtcars, n = -2)
+  expect_equal(res, exp, ignore_attr = TRUE)
```
Maybe just `expect_equal(slice_head(mtcars, n = -2), slice_head(mtcars, n = nrow(mtcars)-2))`

romainfrancois commented on tests/testthat/test-slice.r:
```diff
@@ -189,11 +189,23 @@ test_that("slice() handles matrix and data frame columns (#3630)", {
+  res <- slice_tail(mtcars, n = -2)
+  exp <- tail(mtcars, n = -2)
+  expect_equal(res, exp, ignore_attr = TRUE)
+})
```
This needs a test for grouped data frames as well, e.g. 

``` r
library(dplyr, warn.conflicts = FALSE)
library(testthat)
#> 
#> Attaching package: 'testthat'
#> The following object is masked from 'package:dplyr':
#> 
#>     matches

df <- data.frame(x = 1:10, g = c(rep(1, 8), rep(2, 2))) %>% group_by(g)

expect_equal(
  slice_head(df, n = -3), 
  slice(df, rlang::seq2(1L, n() - 3))
)
expect_equal(
  n_groups(slice_head(df, n = -3)), 
  1L
)
```

<sup>Created on 2021-08-05 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

romainfrancois commented on R/slice.R:
```diff
@@ -306,8 +315,11 @@ check_slice_size <- function(n, prop, .slice_fn = "check_slice_size") {
-      abort("`n` must be a non-missing positive number.")
+    if (is.na(n)) {
+      abort("`n` must be a non-missing number.")
+    }
```
I believe this should go before the previous `if` otherwise it only works for integer NA, which is an issue because `NA` is logical: 

``` r
library(dplyr, warn.conflicts = FALSE)

df <- data.frame(x = 1:10)
slice_head(df, n = NA)
#> Error: `n` must be a single number.
slice_head(df, n = NA_integer_)
#> Error: `n` must be a non-missing number.
```

<sup>Created on 2021-08-05 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

romainfrancois replied:
And then it probably needs a test. Or perhaps, both could be merged into: 

```r
    if (!is.numeric(n) || length(n) != 1 || is.na(n)) {
      abort("`n` must be a single non-missing number.")
    }
```

eutwt replied:
Interesting. This is also the case for `prop` (in both master an this branch). I'm thinking I should make this change for `prop` too then. But, since it's outside the scope of this pull request I can leave it if you prefer.

``` r
library(dplyr, warn.conflicts = FALSE)

df <- data.frame(x = 1:10)
slice_head(df, prop = NA)
#> Error: `prop` must be a single number
slice_head(df, prop = NA_integer_)
#> Error: `prop` must be a non-missing positive number.
```

<sup>Created on 2021-08-05 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

romainfrancois reviewed (changes requested):
I think it's worth supporting negatives, just added a few comments/requests for minor changes

eutwt commented:
All set. Changes are:

- Added test on a grouped data frame for `slice_head` and `slice_tail`
- Corrected the order of validation checks for `n` and `prop` arguments
  - Previously this PR didn't touch `prop` at all, but `prop` has the same behavior as `n` mentioned above, and it seems best to keep the behavior consistent rather than only change one.

romainfrancois reviewed (commented):
LGTM. deferring to @Hadley for approval, i.e. should negative `n` be supported in `slice_head()` / `slice_tail()`

hadley commented on R/slice.R:
```diff
@@ -303,21 +312,21 @@ check_slice_size <- function(n, prop, .slice_fn = "check_slice_size") {
-    if (!is.numeric(n) || length(n) != 1) {
-      abort("`n` must be a single number.")
+    if (!is.numeric(n) || length(n) != 1 || is.na(n)) {
+      abort("`n` must be a single non-missing number.")
```
I don't think adding "non-missing" adds much because IMO "number" implies not missing.

hadley commented on R/slice.R:
```diff
@@ -132,7 +135,10 @@ slice_head.data.frame <- function(.data, ..., n, prop) {
+    n = function(n) {
+      idx_end = if (size$n < 0) max(n + size$n, 0) else min(size$n, n)
+      seq2(1, idx_end)
+    },
```
This changes makes `slice_head()` and `slice_tail()` more consistent with `head()` and `tail()`, which is good, but it makes `n` and `prop` less consistent, which is bad. How about applying the same principle to `prop` as well?

hadley commented on R/slice.R:
```diff
@@ -35,6 +35,9 @@
 #'   proportion of rows to select. If neither are supplied, `n = 1` will be
 #'   used.
 #'
+#"   A positive value of `n` includes the first/last `n` rows, while a negative 
```
Does this imply that the other `slice_` functions also need to be changed?

eutwt replied:
I think that would be a good idea. I can apply this to all variants in this PR if you agree, or just change the wording in the documentation if not.

hadley replied:
Yeah, I think it would make sense to consistently apply to all variants 😄

hadley reviewed (commented):
Seems like a good idea to me 😄. Maybe should also make the connection to `head()` and `tail()` a little stronger in the documentation?

eutwt commented:
@hadley Do you agree with this behavior for non-integer number of rows in the negative case?

```r
test_that("Non-integer number of rows computed correctly", {
  df <- data.frame(x = 1:10)
  
  # Current behavior
  expect_equal(df %>% slice_head(n = 1.6) %>% nrow(), 1)
  expect_equal(df %>% slice_head(prop = 0.16) %>% nrow(), 1) # Since PR 5234
  
  # Negative non-integer. Absolute value rounded down to match positive case
  expect_equal(df %>% slice_head(n = -1.6) %>% nrow(), 9) 
  expect_equal(df %>% slice_head(prop = -0.16) %>% nrow(), 9)
  
})
```

hadley commented:
Yeah, I think the principle should be that `rbind(slice_head(x, -y), slice_head(x, y))` gives you back all the rows (albeit in a different order).

eutwt commented:
Each slice variant had a `switch` statement to branch on whether `n` or `prop` is provided. If I add another negative/positive branch that's four cases to deal with in all five `slice_` variants. The four cases are being dealt with the same way in all variants, so I pulled the calculation of slice size out into a separate function `get_slice_size`. The difference between the slice variants is just the `idx` function's handling of the calculated size.

eutwt commented:
I did add a test of `get_slice_size`, but in addition here's a close to exhaustive comparison of this PR with CRAN dplyr

**Output on CRAN dplyr:**

``` r
library(dplyr, warn.conflicts = FALSE)
df <- data.frame(a = 1:10)

# Positive n
slice_head(df, n = 2.6)
#>   a
#> 1 1
#> 2 2
slice_tail(df, n = 2.6)
#> Error: Can't convert from <double> to <integer> due to loss of precision.
#> * Locations: 1, 2
slice_min(df, n = 2.6, order_by = a)
#>   a
#> 1 1
#> 2 2
slice_max(df, n = 2.6, order_by = a)
#>    a
#> 1 10
#> 2  9
set.seed(2021)
slice_sample(df, n = 2.6)
#>   a
#> 1 7
#> 2 6

# Positive prop
slice_head(df, prop = .26)
#>   a
#> 1 1
#> 2 2
slice_tail(df, prop = .26)
#>    a
#> 1  9
#> 2 10
slice_min(df, prop = .26, order_by = a)
#>   a
#> 1 1
#> 2 2
slice_max(df, prop = .26, order_by = a)
#>    a
#> 1 10
#> 2  9
set.seed(2021)
slice_sample(df, prop = .26)
#>   a
#> 1 7
#> 2 6

# Negative n
slice_head(df, n = -2.6)
#> Error: `n` must be a non-missing positive number.
slice_tail(df, n = -2.6)
#> Error: `n` must be a non-missing positive number.
slice_min(df, n = -2.6, order_by = a)
#> Error: `n` must be a non-missing positive number.
slice_max(df, n = -2.6, order_by = a)
#> Error: `n` must be a non-missing positive number.
set.seed(2021)
slice_sample(df, n = -2.6)
#> Error: `n` must be a non-missing positive number.

# Negative prop
slice_head(df, prop = -.26)
#> Error: `prop` must be a non-missing positive number.
slice_tail(df, prop = -.26)
#> Error: `prop` must be a non-missing positive number.
slice_min(df, prop = -.26, order_by = a)
#> Error: `prop` must be a non-missing positive number.
slice_max(df, prop = -.26, order_by = a)
#> Error: `prop` must be a non-missing positive number.
set.seed(2021)
slice_sample(df, prop = -.26)
#> Error: `prop` must be a non-missing positive number.
```

<sup>Created on 2021-08-16 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Output with this PR:**

``` r
library(dplyr, warn.conflicts = FALSE)
df <- data.frame(a = 1:10)

# Positive n
slice_head(df, n = 2.6)
#>   a
#> 1 1
#> 2 2
slice_tail(df, n = 2.6)
#>    a
#> 1  9
#> 2 10
slice_min(df, n = 2.6, order_by = a)
#>   a
#> 1 1
#> 2 2
slice_max(df, n = 2.6, order_by = a)
#>    a
#> 1 10
#> 2  9
set.seed(2021)
slice_sample(df, n = 2.6)
#>   a
#> 1 7
#> 2 6

# Positive prop
slice_head(df, prop = .26)
#>   a
#> 1 1
#> 2 2
slice_tail(df, prop = .26)
#>    a
#> 1  9
#> 2 10
slice_min(df, prop = .26, order_by = a)
#>   a
#> 1 1
#> 2 2
slice_max(df, prop = .26, order_by = a)
#>    a
#> 1 10
#> 2  9
set.seed(2021)
slice_sample(df, prop = .26)
#>   a
#> 1 7
#> 2 6

# Negative n
slice_head(df, n = -2.6)
#>   a
#> 1 1
#> 2 2
#> 3 3
#> 4 4
#> 5 5
#> 6 6
#> 7 7
#> 8 8
slice_tail(df, n = -2.6)
#>    a
#> 1  3
#> 2  4
#> 3  5
#> 4  6
#> 5  7
#> 6  8
#> 7  9
#> 8 10
slice_min(df, n = -2.6, order_by = a)
#>   a
#> 1 1
#> 2 2
#> 3 3
#> 4 4
#> 5 5
#> 6 6
#> 7 7
#> 8 8
slice_max(df, n = -2.6, order_by = a)
#>    a
#> 1 10
#> 2  9
#> 3  8
#> 4  7
#> 5  6
#> 6  5
#> 7  4
#> 8  3
set.seed(2021)
slice_sample(df, n = -2.6)
#>    a
#> 1  7
#> 2  6
#> 3  9
#> 4  2
#> 5  4
#> 6  8
#> 7 10
#> 8  5

# Positive prop
slice_head(df, prop = -.26)
#>   a
#> 1 1
#> 2 2
#> 3 3
#> 4 4
#> 5 5
#> 6 6
#> 7 7
#> 8 8
slice_tail(df, prop = -.26)
#>    a
#> 1  3
#> 2  4
#> 3  5
#> 4  6
#> 5  7
#> 6  8
#> 7  9
#> 8 10
slice_min(df, prop = -.26, order_by = a)
#>   a
#> 1 1
#> 2 2
#> 3 3
#> 4 4
#> 5 5
#> 6 6
#> 7 7
#> 8 8
slice_max(df, prop = -.26, order_by = a)
#>    a
#> 1 10
#> 2  9
#> 3  8
#> 4  7
#> 5  6
#> 6  5
#> 7  4
#> 8  3
set.seed(2021)
slice_sample(df, prop = -.26)
#>    a
#> 1  7
#> 2  6
#> 3  9
#> 4  2
#> 5  4
#> 6  8
#> 7 10
#> 8  5
```

<sup>Created on 2021-08-16 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

eutwt commented:
> Yeah, I think the principle should be that `rbind(slice_head(x, -y), slice_head(x, y))` gives you back all the rows (albeit in a different order).

I think you probably meant `rbind(slice_head(df, n = -1), slice_tail(df, n = 1))`, otherwise this principle is not true for `utils::head`. I think `head` should only ever give the first part of a data.frame, never the last (and the inverse for `tail`), which precludes the principle as quoted. 

``` r
df <- data.frame(a = 1:3)

head(df, 1)
#>   a
#> 1 1
head(df, -1)
#>   a
#> 1 1
#> 2 2

head(df, 1)
#>   a
#> 1 1
tail(df, -1)
#>   a
#> 2 2
#> 3 3

library(dplyr, warn.conflicts = FALSE) # version from this PR
identical(df, rbind(slice_head(df, n = -1), slice_tail(df, n = 1)))
#> [1] TRUE
```

<sup>Created on 2021-08-16 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Same idea for `min` and `max`, e.g. `min` should always give rows with smaller values, even when `n` or `prop` is negative.

romainfrancois commented:
Thanks @eutwt I'll merge this in now :-)

eutwt commented:
@romainfrancois FYI I forgot to `document()`, so you may see doc-changes not done by you when working on the other slice PRs. Sorry about that
